File: scripts/prevent-main-agent-write.py
#!/usr/bin/env python3
"""Block main-agent writes while allowing registered worker/subagent sessions."""

from __future__ import annotations

import json
import os
import re
import shlex
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Any


SUBAGENT_START_EVENTS = {"SubagentStart", "subagent_start", "subagentStart"}
SUBAGENT_STOP_EVENTS = {"SubagentStop", "subagent_stop", "subagentStop"}
SESSION_START_EVENTS = {"SessionStart", "session_start", "sessionStart"}
PRE_TOOL_USE_EVENTS = {"PreToolUse", "pre_tool_use", "preToolUse", ""}
SHELL_TOOL_NAMES = {"Bash", "shell", "exec", "exec_command", "unified_exec"}
DIRECT_WRITE_TOOL_NAMES = {
  "Write",
  "Edit",
  "MultiEdit",
  "NotebookEdit",
  "apply_patch",
}
SUBAGENT_TOOL_NAMES = {
  "Task",
  "delegate_task",
  "SubagentStart",
  "subagent_start",
  "WorkerStart",
  "worker_start",
}
COMMAND_KEYS = ("command", "cmd")
SESSION_KEYS = ("session_id", "sessionId")
STATE_PATH_ENV = "DEVFLOW_MAIN_AGENT_WRITE_STATE"
STATE_PATH = Path(tempfile.gettempdir()) / f"devflow-main-agent-write-sessions-{os.getuid()}.json"
SHELL_WRITE_COMMANDS = {
  "chmod",
  "chown",
  "cp",
  "dd",
  "install",
  "ln",
  "mkdir",
  "mv",
  "patch",
  "rm",
  "tee",
  "touch",
  "truncate",
}
GIT_WRITE_SUBCOMMANDS = {
  "add",
  "am",
  "apply",
  "cherry-pick",
  "clean",
  "commit",
  "merge",
  "mv",
  "pull",
  "rebase",
  "reset",
  "restore",
  "revert",
  "rm",
  "stash",
}
PACKAGE_MANAGERS = {"npm", "pnpm", "yarn", "bun", "pip", "pip3", "uv", "cargo", "go"}
GIT_GLOBAL_OPTIONS_WITH_VALUE = {"-C", "-c", "--git-dir", "--work-tree", "--namespace"}
GIT_GLOBAL_OPTIONS_WITH_OPTIONAL_VALUE = {"--config-env"}


@dataclass(frozen=True)
class BlockDecision:
  reason: str
  escalation: bool = False


def main() -> int:
  payload = _read_payload()
  if not isinstance(payload, dict):
    return 0

  event = _find_hook_event(payload)
  session_id = _find_session_id(payload)
  if event in SESSION_START_EVENTS:
    return _handle_session_start(payload)
  if event in SUBAGENT_START_EVENTS:
    return _handle_subagent_start(payload, session_id)
  if event in SUBAGENT_STOP_EVENTS:
    return _handle_subagent_stop(session_id)
  if event not in PRE_TOOL_USE_EVENTS:
    return 0

  tool_name = _find_tool_name(payload)
  if tool_name in SUBAGENT_TOOL_NAMES:
    return 0
  tool_input = _find_tool_input(payload) or {}
  decision = should_block_tool(tool_name, tool_input, session_id)
  if decision is None:
    return 0
  _write_block_message(decision)
  return 2


def should_block_tool(
  tool_name: str,
  tool_input: dict[str, Any],
  session_id: str | None,
) -> BlockDecision | None:
  decision = _decision_for_tool(tool_name, tool_input)
  if decision is None:
    return None
  if decision.escalation:
    return decision
  if _is_global_git_push_decision(decision):
    return decision
  if not session_id:
    return BlockDecision(
      f"{decision.reason}，但 payload 缺少 session_id，无法证明这是 worker/subagent 写入"
    )
  if not is_registered_subagent_session(session_id):
    return BlockDecision(f"{decision.reason}，当前 session `{session_id}` 未登记为 worker/subagent")
  return None


def is_registered_subagent_session(session_id: str) -> bool:
  return session_id in _load_state()


def _handle_subagent_start(payload: dict[str, Any], session_id: str | None) -> int:
  if not session_id:
    return 0
  agent_name = _find_agent_name(payload)
  state = _load_state()
  state[session_id] = {"agent": agent_name}
  _save_state(state)
  return 0


def _handle_subagent_stop(session_id: str | None) -> int:
  if not session_id:
    return 0
  state = _load_state()
  if session_id in state:
    state.pop(session_id, None)
    _save_state(state)
  return 0


def _handle_session_start(_payload: dict[str, Any]) -> int:
  sys.stdout.write(
    "\n".join(
      [
        "DevFlow mode: coordinator-only",
        "Main agent may coordinate, review, and verify only.",
        "Worker/subagent sessions may write files.",
        "Read-only inspection commands are allowed.",
      ]
    )
  )
  sys.stdout.write("\n")
  return 0


def _decision_for_tool(tool_name: str, tool_input: dict[str, Any]) -> BlockDecision | None:
  if tool_name in DIRECT_WRITE_TOOL_NAMES:
    return BlockDecision(f"`{tool_name}` 是直接写入工具")
  if tool_name and tool_name not in SHELL_TOOL_NAMES:
    return None
  command = _find_command(tool_input)
  if not command:
    return None
  return _decision_for_command(command)


def _decision_for_command(command: str) -> BlockDecision | None:
  for segment in _command_segments(command):
    proxy_reason = _proxy_escalation_reason(segment)
    if proxy_reason is not None:
      return BlockDecision(proxy_reason, escalation=True)
    normalized = _normalize_command_prefix(segment)
    if not normalized:
      continue
    if _is_test_command(normalized):
      continue
    if _has_shell_redirection(normalized):
      return BlockDecision("shell 重定向会写入文件")
    if normalized[0] == "git":
      reason = _git_push_reason(normalized)
      if reason is not None:
        return BlockDecision(reason)
      reason = _git_write_reason(normalized)
      if reason is not None:
        return BlockDecision(reason)
      continue
    if normalized[0] in PACKAGE_MANAGERS and _package_command_writes(normalized):
      return BlockDecision(f"`{normalized[0]} {' '.join(normalized[1:3])}` 可能修改依赖或锁文件")
    if normalized[0] == "sed":
      reason = _sed_write_reason(normalized)
      if reason is not None:
        return BlockDecision(reason)
      continue
    if normalized[0] in {"python", "python3"}:
      reason = _python_write_reason(normalized)
      if reason is not None:
        return BlockDecision(reason)
      continue
    if normalized[0] in SHELL_WRITE_COMMANDS:
      return BlockDecision(f"`{normalized[0]}` 是写入型 shell 命令")
  return None


def _read_payload() -> Any:
  try:
    raw = sys.stdin.read()
  except OSError:
    return None
  if not raw.strip():
    return None
  try:
    return json.loads(raw)
  except json.JSONDecodeError:
    return None


def _find_hook_event(payload: dict[str, Any]) -> str:
  value = payload.get("hook_event_name") or payload.get("hookEventName") or payload.get("event")
  return value if isinstance(value, str) else ""


def _find_tool_name(payload: dict[str, Any]) -> str:
  value = payload.get("tool_name") or payload.get("toolName") or payload.get("tool")
  return value if isinstance(value, str) else ""


def _find_tool_input(payload: dict[str, Any]) -> dict[str, Any] | None:
  value = payload.get("tool_input") or payload.get("toolInput")
  return value if isinstance(value, dict) else None


def _find_command(tool_input: dict[str, Any]) -> str | None:
  for key in COMMAND_KEYS:
    value = tool_input.get(key)
    if isinstance(value, str) and value.strip():
      return value
  return None


def _find_session_id(payload: dict[str, Any]) -> str | None:
  for key in SESSION_KEYS:
    value = payload.get(key)
    if isinstance(value, str) and value.strip():
      return value
  return None


def _find_agent_name(payload: dict[str, Any]) -> str:
  values = [
    payload.get("agent_type"),
    payload.get("agentType"),
    payload.get("agentName"),
    payload.get("agentDisplayName"),
  ]
  return " ".join(value for value in values if isinstance(value, str)).strip()


def _command_segments(command: str) -> list[list[str]]:
  try:
    lexer = shlex.shlex(command, posix=True, punctuation_chars=";&|<>")
    lexer.whitespace_split = True
    tokens = list(lexer)
  except (TypeError, ValueError):
    tokens = command.split()
  segments: list[list[str]] = []
  current: list[str] = []
  for token in tokens:
    if token in {";", "&&", "||", "|"}:
      if current:
        segments.append(current)
        current = []
      continue
    current.append(token)
  if current:
    segments.append(current)
  return segments


def _normalize_command_prefix(tokens: list[str]) -> list[str]:
  normalized = _strip_launcher_prefix(tokens)
  if normalized and normalized[0] == "rtk":
    normalized.pop(0)
    if normalized and normalized[0] == "proxy":
      normalized.pop(0)
  return normalized


def _proxy_escalation_reason(tokens: list[str]) -> str | None:
  normalized = _strip_launcher_prefix(tokens)
  if normalized and normalized[0] == "proxy":
    return "二级警告：检测到 `proxy` 代理执行，疑似在尝试绕过既有权限或 hook 限制"
  if len(normalized) >= 2 and normalized[0] == "rtk" and normalized[1] == "proxy":
    return "二级警告：检测到 `rtk proxy` 代理执行，疑似在尝试绕过既有权限或 hook 限制"
  return None


def _strip_launcher_prefix(tokens: list[str]) -> list[str]:
  normalized = list(tokens)
  changed = True
  while changed:
    changed = False
    while normalized and _is_env_assignment(normalized[0]):
      normalized.pop(0)
      changed = True
    while normalized and normalized[0] in {"command", "builtin", "exec", "env"}:
      normalized.pop(0)
      changed = True
  return normalized


def _is_env_assignment(token: str) -> bool:
  if "=" not in token:
    return False
  name, _value = token.split("=", 1)
  return bool(name) and re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", name) is not None


def _has_shell_redirection(tokens: list[str]) -> bool:
  return any(
    token in {">", ">>", "1>", "2>", "&>", "<>"} or token.startswith((">", ">>"))
    for token in tokens
  )


def _git_write_reason(tokens: list[str]) -> str | None:
  subcommand = _git_subcommand(tokens[1:])
  if subcommand in GIT_WRITE_SUBCOMMANDS:
    return f"`git {subcommand}` 会修改工作区、索引或提交历史"
  return None


def _git_push_reason(tokens: list[str]) -> str | None:
  subcommand = _git_subcommand(tokens[1:])
  if subcommand == "push":
    return "`git push` 已被全场拦截；Agent 不允许执行任何推送"
  return None


def _git_subcommand(args: list[str]) -> str | None:
  index = 0
  while index < len(args):
    token = args[index]
    if token == "--":
      return args[index + 1] if index + 1 < len(args) else None
    if token in GIT_GLOBAL_OPTIONS_WITH_VALUE:
      index += 2
      continue
    if any(token.startswith(f"{option}=") for option in GIT_GLOBAL_OPTIONS_WITH_VALUE):
      index += 1
      continue
    if token in GIT_GLOBAL_OPTIONS_WITH_OPTIONAL_VALUE:
      index += 2
      continue
    if token.startswith("--") and "=" in token:
      index += 1
      continue
    if token.startswith("-"):
      index += 1
      continue
    return token
  return None


def _is_global_git_push_decision(decision: BlockDecision) -> bool:
  return decision.reason.startswith("`git push` 已被全场拦截")


def _first_non_option(tokens: list[str]) -> str | None:
  index = 0
  while index < len(tokens):
    token = tokens[index]
    if token == "--":
      return tokens[index + 1] if index + 1 < len(tokens) else None
    if token.startswith("-"):
      index += 1
      continue
    return token
  return None


def _package_command_writes(tokens: list[str]) -> bool:
  command = _first_non_option(tokens[1:])
  if command is None:
    return False
  manager = tokens[0]
  if manager in {"npm", "pnpm", "yarn", "bun"}:
    return command in {
      "add",
      "build",
      "generate",
      "codegen",
      "install",
      "i",
      "remove",
      "rm",
      "run",
      "update",
      "upgrade",
    }
  if manager in {"pip", "pip3", "uv"}:
    return command in {"install", "add", "remove", "sync"}
  if manager == "cargo":
    return command in {"add", "remove", "update", "install"}
  if manager == "go":
    return command in {"get", "mod", "work"}
  return False


def _sed_write_reason(tokens: list[str]) -> str | None:
  args = tokens[1:]
  if any(_is_sed_in_place_option(token) for token in args):
    return "`sed -i` 会原地修改文件"
  return None


def _is_sed_in_place_option(token: str) -> bool:
  return token == "-i" or token.startswith("-i") or token == "--in-place" or token.startswith("--in-place=")


def _is_test_command(tokens: list[str]) -> bool:
  command = tokens[0]
  args = tokens[1:]
  if command in {"pytest", "tox", "nox"}:
    return True
  if command in {"npm", "pnpm", "yarn", "bun"}:
    subcommand = _first_non_option(args)
    if subcommand == "test":
      return True
    if subcommand == "run":
      script_name = _script_name_after_run(args)
      return script_name is not None and _is_test_like_script_name(script_name)
    return False
  if command in {"cargo", "go"}:
    return _first_non_option(args) == "test"
  if command in {"mvn", "gradle", "./gradlew"}:
    return any("test" in arg.lower() or "check" in arg.lower() for arg in args)
  if command in {"python", "python3"}:
    return _is_python_safe_test_command(args)
  return False


def _script_name_after_run(args: list[str]) -> str | None:
  run_index = next((index for index, token in enumerate(args) if token == "run"), None)
  if run_index is None:
    return None
  for token in args[run_index + 1 :]:
    if token == "--":
      continue
    if token.startswith("-"):
      continue
    return token
  return None


def _is_test_like_script_name(script_name: str) -> bool:
  normalized = script_name.lower()
  return any(marker in normalized for marker in ("test", "check", "lint", "typecheck"))


def _is_python_safe_test_command(args: list[str]) -> bool:
  filtered = _strip_python_runtime_options(args)
  if not filtered:
    return False
  if filtered[0] == "-m" and len(filtered) >= 2:
    return filtered[1] in {"pytest", "unittest", "py_compile", "json.tool"}
  script_path = Path(filtered[0])
  script = script_path.name
  normalized_path = script_path.as_posix()
  if script.startswith("test-") or script.startswith("test_"):
    return True
  if script.startswith("run_") and script.endswith("examples.py"):
    return True
  return normalized_path == "skills/tdd-skill/scripts/run_protocol_examples.py"


def _python_write_reason(tokens: list[str]) -> str | None:
  filtered = _strip_python_runtime_options(tokens[1:])
  if not filtered:
    return None
  if filtered[0] == "-c":
    code = " ".join(filtered[1:])
    if _python_inline_code_writes(code):
      return "`python -c` 内联代码包含明显文件写入调用"
    return None
  script = Path(filtered[0]).name.lower()
  if any(marker in script for marker in ("write", "update", "generate", "codegen", "modify", "migrate")):
    return f"`python {' '.join(tokens[1:])}` 可能执行写入型脚本"
  return None


def _python_inline_code_writes(code: str) -> bool:
  if re.search(r"open\([^)]*,\s*['\"][^'\"]*[wax+][^'\"]*['\"]", code):
    return True
  return any(
    marker in code
    for marker in (
      ".write(",
      ".write_text(",
      ".write_bytes(",
      "os.remove(",
      "shutil.",
    )
  )


def _strip_python_runtime_options(args: list[str]) -> list[str]:
  filtered = list(args)
  options_with_value = {"-c", "-m", "-W", "-X"}
  result: list[str] = []
  index = 0
  while index < len(filtered):
    token = filtered[index]
    if token == "-m":
      return filtered[index:]
    if token in {"-X", "-W"}:
      index += 2
      continue
    if token.startswith(("-X", "-W")) and len(token) > 2:
      index += 1
      continue
    if token == "-c" or token.startswith("-c"):
      return filtered[index:]
    if token.startswith("-"):
      index += 1
      continue
    result.extend(filtered[index:])
    return result
  return result


def _load_state() -> dict[str, Any]:
  path = _state_path()
  try:
    value = json.loads(path.read_text(encoding="utf-8"))
  except (OSError, json.JSONDecodeError):
    return {}
  return value if isinstance(value, dict) else {}


def _save_state(state: dict[str, Any]) -> None:
  path = _state_path()
  path.parent.mkdir(parents=True, exist_ok=True)
  tmp = path.with_suffix(f"{path.suffix}.tmp")
  tmp.write_text(json.dumps(state, ensure_ascii=False, sort_keys=True), encoding="utf-8")
  os.replace(tmp, path)


def _state_path() -> Path:
  configured = os.environ.get(STATE_PATH_ENV)
  return Path(configured) if configured else STATE_PATH


def _write_block_message(decision: BlockDecision) -> None:
  sys.stderr.write(
    "\n".join(
      [
        "DevFlow 已阻止主 Agent 直接执行写操作。",
        f"原因：{decision.reason}。",
        "",
        "主 Agent 只能协调、审查和验证；请通过 worker/subagent 完成代码写入。",
      ]
    )
  )
  sys.stderr.write("\n")


if __name__ == "__main__":
  raise SystemExit(main())

File: scripts/prevent-main-agent-write.ts
#!/usr/bin/env bun

import {
  readPayload,
  findHookEvent,
  findToolName,
  findToolInput,
  findCommand,
  findWorkdir,
  findToolWorkdir,
  findSessionId,
  findAgentName,
  SESSION_HOOK_NAMES,
  SUBAGENT_START_EVENTS,
  SUBAGENT_STOP_EVENTS,
  PRE_TOOL_USE_EVENTS,
  SHELL_TOOL_NAMES,
  DIRECT_WRITE_TOOL_NAMES,
  SUBAGENT_TOOL_NAMES,
} from "../src/shared/payload"
import {
  commandSegments,
  normalizeCommandPrefix,
  stripLauncherPrefix,
  hasShellRedirection,
  gitSubcommand,
  PACKAGE_MANAGERS,
  SHELL_WRITE_COMMANDS,
  GIT_WRITE_SUBCOMMANDS,
  isPackageCommandWrites,
  isSedWriteCommand,
  pythonWriteReason,
  isTestCommand,
} from "../src/shared/command-parser"
import { currentBranch, PROTECTED_BRANCHES } from "../src/shared/branch"
import { loadState, saveState, isRegisteredSubagentSession, getStatePathEnv } from "../src/shared/state-store"
import { createBlockDecision, type BlockDecision } from "../src/shared/types"
import type { Payload, ToolInput } from "../src/shared/types"

export function shouldBlockTool(
  toolName: string,
  toolInput: ToolInput,
  sessionId?: string,
  cwd?: string,
): BlockDecision | undefined {
  const decision = decisionForTool(toolName, toolInput)
  if (!decision) return undefined
  if (decision.escalation) return decision
  if (isGlobalGitPushDecision(decision)) return decision

  if (!sessionId) {
    return createBlockDecision(
      "unknown",
      `${decision.reason}，但 payload 缺少 session_id，无法证明这是 worker/subagent 写入`,
    )
  }
  if (!isRegisteredSubagentSession(sessionId)) {
    return createBlockDecision(
      "unknown",
      `${decision.reason}，当前 session \`${sessionId}\` 未登记为 worker/subagent`,
    )
  }

  const effectiveCwd = cwd ?? findToolWorkdir(toolInput)
  if (effectiveCwd) {
    const protectedBranchDecision = protectedBranchWriteDecision(effectiveCwd, decision.reason)
    if (protectedBranchDecision) return protectedBranchDecision
  }

  return undefined
}

function decisionForTool(toolName: string, toolInput: ToolInput): BlockDecision | undefined {
  if (DIRECT_WRITE_TOOL_NAMES.has(toolName)) {
    return createBlockDecision("unknown", `\`${toolName}\` 是直接写入工具`)
  }
  if (toolName && !SHELL_TOOL_NAMES.has(toolName)) return undefined

  const command = findCommand(toolInput)
  if (!command) return undefined
  return decisionForCommand(command)
}

function decisionForCommand(command: string): BlockDecision | undefined {
  for (const segment of commandSegments(command)) {
    const proxyReason = proxyEscalationReason(segment)
    if (proxyReason) {
      return createBlockDecision("unknown", proxyReason, "escalation")
    }
    const normalized = normalizeCommandPrefix(segment)
    if (!normalized.length) continue

    if (isTestCommand(normalized)) continue

    if (hasShellRedirection(normalized)) {
      return createBlockDecision("unknown", "shell 重定向会写入文件")
    }
    if (normalized[0] === "git") {
      const pushReason = gitPushReason(normalized)
      if (pushReason) return createBlockDecision("unknown", pushReason)
      const writeReason = gitWriteReason(normalized)
      if (writeReason) return createBlockDecision("unknown", writeReason)
      continue
    }
    if (PACKAGE_MANAGERS.has(normalized[0]) && isPackageCommandWrites(normalized)) {
      const packageName = normalized.slice(0, 3).join(" ")
      return createBlockDecision("unknown", `\`${packageName}\` 可能修改依赖或锁文件`)
    }
    if (normalized[0] === "sed") {
      const reason = isSedWriteCommand(normalized)
      if (reason) return createBlockDecision("unknown", "`sed -i` 会原地修改文件")
      continue
    }
    if (normalized[0] === "python" || normalized[0] === "python3") {
      const reason = pythonWriteReason(normalized.slice(1))
      if (reason) return createBlockDecision("unknown", reason)
      continue
    }
    if (SHELL_WRITE_COMMANDS.has(normalized[0])) {
      return createBlockDecision("unknown", `\`${normalized[0]}\` 是写入型 shell 命令`)
    }
  }
  return undefined
}

function proxyEscalationReason(tokens: string[]): string | undefined {
  const normalized = stripLauncherPrefix(tokens)
  if (normalized.length && normalized[0] === "proxy") {
    return "二级警告：检测到 `proxy` 代理执行，疑似在尝试绕过既有权限或 hook 限制"
  }
  if (normalized.length >= 2 && normalized[0] === "rtk" && (normalized[1] as string) === "proxy") {
    return "二级警告：检测到 `rtk proxy` 代理执行，疑似在尝试绕过既有权限或 hook 限制"
  }
  return undefined
}

function gitPushReason(tokens: string[]): string | undefined {
  const subcommand = gitSubcommand(tokens.slice(1))
  if (subcommand === "push") {
    return "`git push` 已被全场拦截；Agent 不允许执行任何推送"
  }
  return undefined
}

function gitWriteReason(tokens: string[]): string | undefined {
  const subcommand = gitSubcommand(tokens.slice(1))
  if (subcommand && GIT_WRITE_SUBCOMMANDS.has(subcommand)) {
    return `\`git ${subcommand}\` 会修改工作区、索引或提交历史`
  }
  return undefined
}

function isGlobalGitPushDecision(decision: BlockDecision): boolean {
  return decision.reason.startsWith("`git push` 已被全场拦截")
}

function protectedBranchWriteDecision(cwd: string, reason: string): BlockDecision | undefined {
  const branch = currentBranch(cwd)
  if (branch && PROTECTED_BRANCHES.has(branch)) {
    return createBlockDecision("unknown", `${reason}，当前分支 \`${branch}\` 受保护（cwd: ${cwd}）`)
  }
  return undefined
}

function handleSubagentStart(payload: Payload, sessionId: string | undefined): number {
  if (!sessionId) return 0
  const agentName = findAgentName(payload)
  const state = loadState()
  state[sessionId] = { agent: agentName }
  saveState(state)
  return 0
}

function handleSubagentStop(sessionId: string | undefined): number {
  if (!sessionId) return 0
  const state = loadState()
  if (sessionId in state) {
    delete state[sessionId]
    saveState(state)
  }
  return 0
}

function handleSessionStart(): number {
  const lines = [
    "DevFlow mode: coordinator-only",
    "Main agent may coordinate, review, and verify only.",
    "Worker/subagent sessions may write files.",
    "Read-only inspection commands are allowed.",
  ]
  for (const line of lines) {
    process.stdout.write(line + "\n")
  }
  return 0
}

function writeBlockMessage(decision: BlockDecision): void {
  const lines = [
    "DevFlow 已阻止主 Agent 直接执行写操作。",
    `原因：${decision.reason}。`,
    "",
    "主 Agent 只能协调、审查和验证；请通过 worker/subagent 完成代码写入。",
  ]
  for (const line of lines) {
    process.stderr.write(line + "\n")
  }
}

function main(): number {
  const payload = readPayload()
  if (!payload || typeof payload !== "object") return 0

  const event = findHookEvent(payload)
  const sessionId = findSessionId(payload)

  if (SESSION_HOOK_NAMES.has(event)) {
    return handleSessionStart()
  }
  if (SUBAGENT_START_EVENTS.has(event)) {
    return handleSubagentStart(payload, sessionId)
  }
  if (SUBAGENT_STOP_EVENTS.has(event)) {
    return handleSubagentStop(sessionId)
  }
  if (!PRE_TOOL_USE_EVENTS.has(event)) {
    return 0
  }

  const toolName = findToolName(payload)
  if (SUBAGENT_TOOL_NAMES.has(toolName)) return 0

  const toolInput = findToolInput(payload) ?? {}
  const cwd = findWorkdir(payload, toolInput)
  const decision = shouldBlockTool(toolName, toolInput, sessionId, cwd)

  if (decision) {
    writeBlockMessage(decision)
    return 2
  }
  return 0
}

if (import.meta.main) {
  process.exit(main())
}
